File: lib/features/clan/models/clan_war_log.dart
import 'dart:convert';
import 'package:clashkingapp/features/clan/models/clan_badge.dart';
import 'package:http/http.dart' as http;

class ClanWarLog {
  final List<WarLogDetails> items;
  final String clanTag;
  late WarLogStats warLogStats;

  ClanWarLog({required this.items, required this.clanTag});

  factory ClanWarLog.fromJson(Map<String, dynamic> json, String clanTag) {
    var itemList = json['items'] != null
        ? (json['items'] as List)
            .where((itemJson) {
              // Extract the endTime and parse it to DateTime
              DateTime endTime = DateTime.parse(itemJson['endTime']);
              // Keep items where the endTime is in 2022 or later
              return endTime.year >= 2022;
            })
            .map((itemJson) =>
                WarLogDetails.fromJson(itemJson as Map<String, dynamic>, clanTag))
            .toList()
        : [];
    return ClanWarLog(items: itemList.cast<WarLogDetails>(), clanTag: clanTag);
  }
}


class WarLogStats {
  final int totalWins;
  final int totalLosses;
  final int totalTies;
  final int totalWars;
  final int averageMembers; 
  final double averageClanDestruction;
  final double averageClanStarsPerMember;
  final double averageOpponentDestruction;
  final double averageOpponentStarsPerMember;
  final double averageAttacksPerMember;
  final String winPercentage;
  final String lossPercentage;
  final String tiePercentage;
  final double averageDestructionDifference;
  final double averageClanStarsPercentage;
  final double averageOpponentStarsPercentage;

  WarLogStats({
    required this.totalWins,
    required this.totalLosses,
    required this.totalTies,
    required this.totalWars,
    required this.averageMembers,
    required this.averageClanDestruction,
    required this.averageClanStarsPerMember,
    required this.averageOpponentDestruction,
    required this.averageOpponentStarsPerMember,
    required this.averageAttacksPerMember,
    required this.winPercentage,
    required this.lossPercentage,
    required this.tiePercentage,
    required this.averageDestructionDifference,
    required this.averageClanStarsPercentage,
    required this.averageOpponentStarsPercentage,
  });

  // toString method to print the object as a string
  @override
  String toString() {
    return 'WarLogStats: {'
        'totalWins: $totalWins, '
        'totalLosses: $totalLosses, '
        'totalTies: $totalTies, '
        'totalWars: $totalWars, '
        'averageMembers: $averageMembers, '
        'averageClanDestruction: $averageClanDestruction, '
        'averageClanStarsPerMember: $averageClanStarsPerMember, '
        'averageOpponentDestruction: $averageOpponentDestruction, '
        'averageOpponentStarsPerMember: $averageOpponentStarsPerMember, '
        'averageAttacksPerMember: $averageAttacksPerMember, '
        'winPercentage: $winPercentage, '
        'lossPercentage: $lossPercentage, '
        'tiePercentage: $tiePercentage, '
        'averageDestructionDifference: $averageDestructionDifference, '
        'averageClanStarsPercentage: $averageClanStarsPercentage, '
        'averageOpponentStarsPercentage: $averageOpponentStarsPercentage'
        '}';
  }

}

class WarLogDetails {
  final String result;
  final String clanTag;
  final DateTime endTime;
  final int teamSize;
  final int attacksPerMember;
  final ClanDetails clan;
  final ClanDetails opponent;

  WarLogDetails({
    required this.result,
    required this.clanTag,
    required this.endTime,
    required this.teamSize,
    required this.attacksPerMember,
    required this.clan,
    required this.opponent,
  });

  factory WarLogDetails.fromJson(Map<String, dynamic> json, String clanTag) {
    return WarLogDetails(
      result: json['result'] ?? '',
      endTime: DateTime.parse(json['endTime']),
      teamSize: json['teamSize'] ?? 0,
      attacksPerMember: json['attacksPerMember'] ?? 1,
      clan: ClanDetails.fromJson(json['clan']),
      opponent: ClanDetails.fromJson(json['opponent']),
      clanTag: clanTag,
    );
  }
}

class ClanDetails {
  final String tag;
  final String name;
  final ClanBadgeUrls badgeUrls;
  final int clanLevel;
  final int attacks;
  final int stars;
  final double destructionPercentage;
  final int expEarned;

  ClanDetails({
    required this.tag,
    required this.name,
    required this.badgeUrls,
    required this.clanLevel,
    required this.attacks,
    required this.stars,
    required this.destructionPercentage,
    required this.expEarned,
  });

  factory ClanDetails.fromJson(Map<String, dynamic> json) {
    return ClanDetails(
      tag: json['tag'] ?? '',
      name: json['name'] ?? '',
      badgeUrls: ClanBadgeUrls.fromJson(json['badgeUrls'] ?? {}),
      clanLevel: json['clanLevel'] ?? 0,
      attacks: json['attacks'] ?? 0,
      stars: json['stars'] ?? 0,
      destructionPercentage:
          (json['destructionPercentage'] as num?)?.toDouble() ?? 0.0,
      expEarned: json['expEarned'] ?? 0,
    );
  }
}

class WarLogService {
  static Future<ClanWarLog> fetchWarLogData(String tag) async {
    final response = await http.get(Uri.parse(
        'https://proxy.clashk.ing/v1/clans/${tag.replaceAll('#', '%23')}/warlog'));

    if (response.statusCode == 200) {
      String body = utf8.decode(response.bodyBytes);
      Map<String, dynamic> jsonBody = json.decode(body);
      ClanWarLog warLog = ClanWarLog.fromJson(jsonBody, tag);
      warLog.warLogStats =
          await WarLogStatsService.analyzeWarLogs(warLog.items);
      return warLog;
    } else if (response.statusCode == 403) {
      return ClanWarLog(items: [], clanTag: tag);
    } else {
      throw Exception('Failed to load war history data');
    }
  }
}

class WarLogStatsService {
  static Future<WarLogStats> analyzeWarLogs(List<WarLogDetails> warLogs) async {
    int totalWins = 0;
    int totalLosses = 0;
    int totalTies = 0;
    int totalWars = 0;
    int totalMembers = 0;
    int totalAttacks = 0;
    double clanTotalDestruction = 0;
    int clanTotalStars = 0;
    double opponentTotalDestruction = 0;
    int opponentTotalStars = 0;
    int maxPossibleStars = 0;

    for (var log in warLogs) {
      if (log.attacksPerMember == 2) {
        totalWars++;
        switch (log.result) {
          case 'win':
            totalWins++;
            break;
          case 'lose':
            totalLosses++;
            break;
          case 'tie':
            totalTies++;
            break;
        }
        int possibleStarsForThisWar = log.teamSize * 3;
        maxPossibleStars += possibleStarsForThisWar;
        totalMembers += log.teamSize;
        totalAttacks += log.clan.attacks;
        clanTotalDestruction += log.clan.destructionPercentage;
        clanTotalStars += log.clan.stars;
        opponentTotalDestruction += log.opponent.destructionPercentage;
        opponentTotalStars += log.opponent.stars;
      }
    }

    double averageMembers = totalWars > 0 ? totalMembers / totalWars : 0;
    double averageClanDestruction =
        totalWars > 0 ? clanTotalDestruction / totalWars : 0;
    double averageClanStarsPerMember =
        totalMembers > 0 ? clanTotalStars / totalMembers : 0;
    double averageOpponentDestruction =
        totalWars > 0 ? opponentTotalDestruction / totalWars : 0;
    double averageOpponentStarsPerMember =
        totalMembers > 0 ? opponentTotalStars / totalMembers : 0;
    double averageClanStarsPercentage =
        maxPossibleStars > 0 ? (clanTotalStars / maxPossibleStars) * 100 : 0;
    double averageOpponentStarsPercentage =
        maxPossibleStars > 0 ? (opponentTotalStars / maxPossibleStars) * 100 : 0;
    double averageAttacksPerMember =
        totalMembers > 0 ? totalAttacks / totalMembers : 0;
    double winPercentage = totalWars > 0 ? (totalWins / totalWars) * 100 : 0;
    double lossPercentage = totalWars > 0 ? (totalLosses / totalWars) * 100 : 0;
    double tiePercentage = totalWars > 0 ? (totalTies / totalWars) * 100 : 0;
    double averageDestructionDifference =
        averageClanDestruction - averageOpponentDestruction;

    return WarLogStats(
      totalWins: totalWins,
      totalLosses: totalLosses,
      totalTies: totalTies,
      totalWars: totalWars,
      averageMembers: int.parse(averageMembers.toStringAsFixed(0)),
      averageClanDestruction:
          double.parse(averageClanDestruction.toStringAsFixed(0)),
      averageClanStarsPerMember:
          double.parse(averageClanStarsPerMember.toStringAsFixed(1)),
      averageOpponentDestruction:
          double.parse(averageOpponentDestruction.toStringAsFixed(0)),
      averageOpponentStarsPerMember:
          double.parse(averageOpponentStarsPerMember.toStringAsFixed(1)),
      averageClanStarsPercentage:
          double.parse(averageClanStarsPercentage.toStringAsFixed(1)),
      averageOpponentStarsPercentage:
          double.parse(averageOpponentStarsPercentage.toStringAsFixed(1)),
      averageAttacksPerMember:
          double.parse(averageAttacksPerMember.toStringAsFixed(1)),
      winPercentage: winPercentage.toStringAsFixed(0),
      lossPercentage: lossPercentage.toStringAsFixed(0),
      tiePercentage: tiePercentage.toStringAsFixed(0),
      averageDestructionDifference:
          double.parse(averageDestructionDifference.toStringAsFixed(1)),
    );
  }
}

File: expo/src/features/clan/models/clan-war-log.ts
import { ApiClient } from '../../../core/api/client';
import { WarInfoSnapshot } from '../../player/models';
import { ClanBadgeUrls } from './clan-core';
import {
  apiDateOrEpoch,
  int,
  number,
  record,
  records,
  roundTo,
  string,
  type JsonRecord,
} from './parsing';

export class ClanDetails {
  constructor(
    readonly tag: string,
    readonly name: string,
    readonly badgeUrls: ClanBadgeUrls,
    readonly clanLevel: number,
    readonly attacks: number,
    readonly stars: number,
    readonly destructionPercentage: number,
    readonly expEarned: number,
  ) {}
  static fromJson(json: JsonRecord) {
    return new ClanDetails(
      string(json.tag),
      string(json.name),
      ClanBadgeUrls.fromJson(json.badgeUrls),
      int(json.clanLevel),
      int(json.attacks),
      int(json.stars),
      number(json.destructionPercentage),
      int(json.expEarned),
    );
  }
}

export class WarLogDetails {
  constructor(
    readonly result: string,
    readonly clanTag: string,
    readonly endTime: Date,
    readonly teamSize: number,
    readonly attacksPerMember: number,
    readonly clan: ClanDetails,
    readonly opponent: ClanDetails,
  ) {}
  static fromJson(json: JsonRecord, clanTag: string) {
    return new WarLogDetails(
      string(json.result),
      clanTag,
      apiDateOrEpoch(json.endTime),
      int(json.teamSize),
      int(json.attacksPerMember, 1),
      ClanDetails.fromJson(record(json.clan)),
      ClanDetails.fromJson(record(json.opponent)),
    );
  }
}

export class WarLogStats {
  constructor(
    readonly totalWins: number,
    readonly totalLosses: number,
    readonly totalTies: number,
    readonly totalWars: number,
    readonly averageMembers: number,
    readonly averageClanDestruction: number,
    readonly averageClanStarsPerMember: number,
    readonly averageOpponentDestruction: number,
    readonly averageOpponentStarsPerMember: number,
    readonly averageAttacksPerMember: number,
    readonly winPercentage: string,
    readonly lossPercentage: string,
    readonly tiePercentage: string,
    readonly averageDestructionDifference: number,
    readonly averageClanStarsPercentage: number,
    readonly averageOpponentStarsPercentage: number,
  ) {}

  static empty() {
    return new WarLogStats(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '0', '0', '0', 0, 0, 0);
  }
}

export class ClanWarLog {
  private stats: WarLogStats | null = null;

  constructor(
    readonly items: readonly WarLogDetails[],
    readonly clanTag: string,
    readonly wars: readonly WarInfoSnapshot[] = [],
    readonly isPrivate = false,
    readonly reconstructed = false,
  ) {}

  static fromJson(json: JsonRecord, clanTag: string): ClanWarLog {
    const rawItems = records(json.items);
    const items = rawItems
      .filter((item) => strictWarDate(item.endTime).getUTCFullYear() >= 2022)
      .map((item) => WarLogDetails.fromJson(item, clanTag));
    const wars = rawItems.map((item) =>
      WarInfoSnapshot.fromJson({
        ...item,
        state: 'warEnded',
        warType: item.warType ?? (string(item.result) ? 'random' : 'friendly'),
      }),
    );
    return new ClanWarLog(
      items,
      clanTag,
      wars,
      json.isPrivate === true,
      json.reconstructed === true,
    );
  }

  get warLogStats(): WarLogStats {
    return this.stats ?? WarLogStats.empty();
  }

  set warLogStats(value: WarLogStats) {
    this.stats = value;
  }
}

export function analyzeWarLogs(warLogs: readonly WarLogDetails[]): WarLogStats {
  let totalWins = 0;
  let totalLosses = 0;
  let totalTies = 0;
  let totalWars = 0;
  let totalMembers = 0;
  let totalAttacks = 0;
  let clanTotalDestruction = 0;
  let clanTotalStars = 0;
  let opponentTotalDestruction = 0;
  let opponentTotalStars = 0;
  let maxPossibleStars = 0;

  for (const log of warLogs) {
    if (log.attacksPerMember !== 2) continue;
    totalWars += 1;
    if (log.result === 'win') totalWins += 1;
    if (log.result === 'lose') totalLosses += 1;
    if (log.result === 'tie') totalTies += 1;
    maxPossibleStars += log.teamSize * 3;
    totalMembers += log.teamSize;
    totalAttacks += log.clan.attacks;
    clanTotalDestruction += log.clan.destructionPercentage;
    clanTotalStars += log.clan.stars;
    opponentTotalDestruction += log.opponent.destructionPercentage;
    opponentTotalStars += log.opponent.stars;
  }

  const averageMembers = totalWars ? totalMembers / totalWars : 0;
  const averageClanDestruction = totalWars ? clanTotalDestruction / totalWars : 0;
  const averageOpponentDestruction = totalWars ? opponentTotalDestruction / totalWars : 0;
  return new WarLogStats(
    totalWins,
    totalLosses,
    totalTies,
    totalWars,
    Math.round(averageMembers),
    roundTo(averageClanDestruction, 0),
    roundTo(totalMembers ? clanTotalStars / totalMembers : 0, 1),
    roundTo(averageOpponentDestruction, 0),
    roundTo(totalMembers ? opponentTotalStars / totalMembers : 0, 1),
    roundTo(totalMembers ? totalAttacks / totalMembers : 0, 1),
    (totalWars ? (totalWins / totalWars) * 100 : 0).toFixed(0),
    (totalWars ? (totalLosses / totalWars) * 100 : 0).toFixed(0),
    (totalWars ? (totalTies / totalWars) * 100 : 0).toFixed(0),
    roundTo(averageClanDestruction - averageOpponentDestruction, 1),
    roundTo(maxPossibleStars ? (clanTotalStars / maxPossibleStars) * 100 : 0, 1),
    roundTo(maxPossibleStars ? (opponentTotalStars / maxPossibleStars) * 100 : 0, 1),
  );
}

export class WarLogStatsService {
  static analyzeWarLogs(warLogs: readonly WarLogDetails[]): Promise<WarLogStats> {
    return Promise.resolve(analyzeWarLogs(warLogs));
  }
}

export class WarLogService {
  static async fetchWarLogData(
    api: ApiClient,
    tag: string,
    options: { isWarLogPublic: boolean },
  ): Promise<ClanWarLog> {
    const endpoint = options.isWarLogPublic
      ? `/clans/${encodeURIComponent(tag)}/warlog?limit=50`
      : `/clan/${encodeURIComponent(tag)}/warlog?limit=50`;
    const response = options.isWarLogPublic
      ? await api.proxyGet(endpoint)
      : await api.get(endpoint, { requiresAuth: true });
    const parsed: unknown = JSON.parse(response.bodyText);
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed))
      throw new TypeError('Invalid war log response.');
    const result = ClanWarLog.fromJson(parsed as JsonRecord, tag);
    result.warLogStats = analyzeWarLogs(result.items);
    return result;
  }
}

function strictWarDate(value: unknown): Date {
  const date = apiDateOrEpoch(value);
  if (date.getTime() === 0 && String(value) !== '1970-01-01T00:00:00.000Z')
    throw new RangeError(`Invalid war endTime: ${String(value)}`);
  return date;
}
